adapters/mock-da/src/service.rs

```diff
@@ -79,6 +79,62 @@ impl MockDaService {
         }
         anyhow::bail!("No blob at height={height} has been sent in time")
     }
+
+    async fn add_blob(&self, blob: &[u8], zkp_proof: Vec<u8>) -> Result<u64, anyhow::Error> {
+        let mut blocks = self.blocks.write().await;
+
+        let (previous_block_hash, height) = match blocks.iter().last().map(|b| b.header().clone()) {
+            None => (MockHash::from([0; 32]), 0),
+            Some(block_header) => (block_header.hash(), block_header.height + 1),
+        };
+        let data_hash = hash_to_array(blob);
+        let proof_hash = hash_to_array(&zkp_proof);
+        // Hash only from single blob
+        let block_hash = block_hash(height, data_hash, proof_hash, previous_block_hash.into());
+
+        let blob = MockBlob::new_with_zkp_proof(
+            blob.to_vec(),
+            zkp_proof,
+            self.sequencer_da_address,
+            data_hash,
+        );
+        let header = MockBlockHeader {
+            prev_hash: previous_block_hash,
+            hash: block_hash,
+            height,
+            time: Time::now(),
+        };
+        let block = MockBlock {
+            header,
+            validity_cond: Default::default(),
+            blobs: vec![blob],
+        };
+        blocks.push_back(block);
+
+        // Enough blocks to finalize block
+        if blocks.len() > self.blocks_to_finality as usize {
+            let oldest_available_height = blocks[0].header().height();
+            let last_finalized_height = self.last_finalized_height.load(Ordering::Acquire);
```

Comment:
The operations we perform between the `Acquire` and `Release` are relatively complex, including a `send.finalized_header_sender.send` that will involve external resources.

Consider using a lock guard (RwLock/Mutex) instead of atomic synchronization, as the latter is expected to have very cheap ordered operations. This way, you will also unlock the ability to return logical errors instead of panic (i.e. `checked_sub`, `last_finalized_index`, `send.unwrap()`)

Reply:
I moved the code from another place and didn't change the functionality. In line `87`, we are holding a lock for the entire function so I am not sure why we need the `Ordering` here. I will create a separate issue to investigate it.

Reply:
Sounds good! I think the lock does make sense as we manipulate the value that is outside the `self.blocks` lock. But OTOH we are facing this issue where the operation inside the atomic lock is just too complex, so maybe we want to lock a more complex structure than a single `VecDeque` for the blocks? A wrapper `struct BlockState { blocks: VecDeque<Block>, last_finalized_height: u64 }` might make more sense